feat: Add /home API for user stats

haroldadmin opened the pull request:
This commit adds a /home endpoint which returns statistics regarding the
current user. Fixes Issue #120. Tests have been added as well.

### Description
I have added an endpoint in the User Resource. This endpoint triggers a scan of all the relations of the user, and counts the number of these relations in all of the possible states. It then returns a dictionary containing these counted values

Fixes #120 

### Type of Change:
**Delete irrelevant options.**

- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)



### How Has This Been Tested?
I have tested the API using postman, and created my own unit tests as well.


### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

haroldadmin commented:
I have now added all the required changes. The /home endpoint now returns the following details:
1. Number of requests and relations
2. User's name which will used to display in the welcome text on the Android app homescreen
3. A max of 3 recent achievements of the user

I have also added tests for this endpoint. Please review soon :)
If this PR gets merged, then I would also like to work on using this API endpoint in the Android app to create the homescreen (if the issue is still available)

EDIT: Whoops, been spotting a lot of last minute issues, sorry for this messy PR. It's all done now, hopefully.

ramitsawhney27 commented on app/api/dao/user.py:
```diff
@@ -188,3 +189,53 @@ def authenticate(username_or_email, password):
+        tasks = []
+        for relation in all_relations:
+            tasks += relation.tasks_list.tasks
+        achievements = list(filter(lambda x: x.get("is_done"), tasks))
```
Better to use list comprehensions over lambdas and map/reduce/filter.

isabelcosta replied:
@haroldadmin can you please fix this line according to @ramitsawhney27 's suggestion? I had to do a [quick search](https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Comprehensions.html), but this seems to make sense.

haroldadmin replied:
Okay, I'll change this to a list comprehension.

isabelcosta commented on app/api/dao/user.py:
```diff
@@ -188,3 +189,53 @@ def authenticate(username_or_email, password):
+        if achievements:
+            # We only need the first three of these achievements
+            achievements = achievements[0:3]
+            sorted(achievements, key=lambda x: x.get("created_at"))
```
Can you avoid lambdas here as well?

haroldadmin replied:
I think a lambda here is more readable than say, a cmp. However, if the lambda has to be removed then `itemgetter` from `operator` can be used instead. I'll replace the lambda with that.

isabelcosta commented on app/api/dao/user.py:
```diff
@@ -188,3 +189,53 @@ def authenticate(username_or_email, password):
+        return response
+
+    @staticmethod
+    def verify_user(user_id):
```
why is this method here? Were you testing something? Please remove this, since its being done by another function

haroldadmin replied:
I'm sorry, this should not have been in there. Yes, I was testing something. I'll remove it.

isabelcosta commented on app/api/resources/user.py:
```diff
@@ -255,3 +255,18 @@ def post(cls):
         }, 200
 
+
+# TODO: Add swagger docs
```
remove this "todo"

haroldadmin replied:
Okay, will remove this.

isabelcosta commented on app/api/resources/user.py:
```diff
@@ -255,3 +255,18 @@ def post(cls):
+        """
+        user_id = get_jwt_identity()
+        stats = DAO.get_user_statistics(user_id)
+        return marshal(stats, home_response_body_model)
```
Check what response is this API responding in case of success please, because this should return 200 as HTTP response. I tend to put aside with the marshal, but maybe its not necessary

isabelcosta reviewed (changes requested):
@haroldadmin you did a great job with this PR. I just have some small change requests. Can you do the following:
- create tests for DAO functions you created (the API tests are really complete)
- add some doc strings to DAO functions according to [this style guide for doc strings in python by Google](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings)
- Don't forget to squash commits at the end.

haroldadmin commented:
@isabelcosta @ramitsawhney27 I have added the requested changes in the latest commit.

isabelcosta commented on app/api/dao/user.py:
```diff
@@ -188,3 +190,79 @@ def authenticate(username_or_email, password):
+
+    @staticmethod
+    def get_user_statistics(user_id):
+        """
```
In front of `"""` you can have a small summary of the function

You can add this:
```
"""Shows some basic user statistics
```

isabelcosta commented on app/api/dao/user.py:
```diff
@@ -188,3 +190,79 @@ def authenticate(username_or_email, password):
+
+    @staticmethod
+    def get_achievements(user_id):
+        """
```
In front of `"""` you can have a small summary of the function

You can add this:
```
"""Shows a subset of the user's achievements
```

haroldadmin replied:
I wrote it this way initially following the Google style guidelines you linked me to earlier, but then I saw that most of the mentorship-backend project followd the style where the docstring started after the `"""` line. So I figured it is better to follow the style of the project I'm working on.

I've changed the docstring to reflect the style you have suggested.

isabelcosta commented on app/api/resources/user.py:
```diff
@@ -255,3 +255,25 @@ def post(cls):
+    @jwt_required
+    @users_ns.expect(auth_header_parser)
+    def get(cls):
+        """
```
Move the line below next to the `"""`, and like this:
```
"""Get statistics from current user
```

haroldadmin replied:
Yup, changed this too now.

isabelcosta commented on tests/users/test_api_home_statistics.py:
```diff
@@ -0,0 +1,246 @@
+            'achievements': []
+        }
+        auth_header = get_test_request_header(self.user1.id)
+        actual_response = self.client.get('/home', follow_redirects=True, headers=auth_header)
```
Can you also test the status code here?
```
 self.assertEqual(200, actual_response.status_code)
```

And wherever you test a call to the API?

haroldadmin replied:
Good idea, I've now also added this test to all the API call tests. Thanks.

isabelcosta reviewed (changes requested):
Thank you for addressing the changes!

I requested some changes to the doc strings and one little thing in the API tests.
Once you address them I'll wait some time to see if there are other views on this PR and then I can merge this :)

haroldadmin commented:
All requested changes added. Thank you for helping me @isabelcosta 
For the failing codacy check, I've just removed the named arguments syntax from the DAO functions. This prevents the appearance of the word 'password' in there. This change should hopefully make the check pass.

@ramitsawhney27 @m-murad Your feedback on this PR would be very welcome.

isabelcosta commented:
I approved this, I will wait for other PRs on the backend to be ready for merge so then I can merge more than one PR, since these will reset the current DB

isabelcosta commented on tests/users/test_api_home_statistics.py:
```diff
@@ -0,0 +1,253 @@
+
+    def test_completed_relations(self):
+        start_date = datetime.now()
+
```
remove this blank line

isabelcosta commented on tests/users/test_api_home_statistics.py:
```diff
@@ -0,0 +1,253 @@
+
+    def test_accepted_requests_auth(self):
+        start_date = datetime.now()
+
```
remove this blank line

isabelcosta commented on tests/users/test_api_home_statistics.py:
```diff
@@ -0,0 +1,253 @@
+
+    def test_pending_requests_auth(self):
+        start_date = datetime.now()
+
```
remove this blank line

isabelcosta commented:
@haroldadmin left some comments from unnecessary blank lines you had. Could you please remove those?

haroldadmin commented:
@isabelcosta PEP8 style guidelines for Python recommend having a blank line at the end of the file, so I added them there.

The reason is that many text parsing tools like grep rely on a newline character to demarcate the end of a line. You can find more details in the discussion [on this page](https://coderwall.com/p/icxxna/don-t-forget-that-extra-line).

isabelcosta commented:
@haroldadmin I asked you to remove blank lines you have in the middle of the file, not in the end of the file. I agree with the ones in the end of the file, you some in the middle of the test definitions

haroldadmin commented:
Oh, got it. I'll remove them.

haroldadmin commented:
...And it's done! I've added the changes @isabelcosta

EDIT: I don't know what's going on with Travis. Both the failing tests on Travis pass on my system. I ran the tests before pushing the commit. Also, the Travis error is a very weird one: name datetime is not defined, but clearly datetime is imported at that the top.

I'll try pushing the same commit to trigger another build, maybe it will give me more information to work with.

haroldadmin committed:
feat: Add /home API for user stats

This commit adds a /home endpoint which returns statistics regarding the
current user. Fixes Issue #120. Also adds tests for these changes.

haroldadmin commented:
@isabelcosta All done now, hopefully. Please let me know if any further changes are required.

isabelcosta reviewed (approved):
Good work @haroldadmin I don't have anything else to point out.

haroldadmin commented:
@isabelcosta Can this be merged now? I want to work on the Homescreen for the Android app, but this PR needs to be merged for that.

isabelcosta commented:
@haroldadmin later tonight (my timezone) I can merge yes. I was waiting for a PR to be ready for merge. But I'll move on with this one and another one.

isabelcosta commented:
@haroldadmin this PR was merged, but you did something here that flask-restplus does not understand, so the Swagger UI is not being generated. I think we don't need to revert this PR, because it still works on the Android client, but this has to be fixed for people who use the backend. I'll take a look at this, but if you could do that too it would be great.

As you can see here (http://systers-mentorship-dev.eu-central-1.elasticbeanstalk.com/) Swagger UI does not show up.

When adding an API, always to test it on a browser going to http://localhost:5000 as see if the Swagger UI is generated.
